src/Processors/QueryPlan/SortingStep.cpp

```diff
@@ -111,129 +114,174 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
+void SortingStep::transformPipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
+{
+    /// We consider that a caller has more information what type of sorting to apply.
+    /// The type depends on constructor used to create sorting step.
+    /// So we'll try to infer sorting to use only in case of Full sorting
 
-            return std::make_shared<MergeSortingTransform>(
-                    header, result_description, max_block_size, limit, increase_sort_description_compile_attempts_current,
-                    max_bytes_before_remerge / pipeline.getNumStreams(),
-                    remerge_lowered_memory_bytes_ratio,
-                    max_bytes_before_external_sort,
-                    tmp_volume,
-                    min_free_disk_space);
-        });
+    if (type == Type::MergingSorted)
+    {
+        mergingSorted(pipeline, result_description, limit);
+        return;
+    }
 
-        /// If there are several streams, then we merge them into one
-        if (pipeline.getNumStreams() > 1)
+    if (type == Type::FinishSorting)
+    {
+        bool need_finish_sorting = (prefix_description.size() < result_description.size());
+        mergingSorted(pipeline, prefix_description, (need_finish_sorting ? 0 : limit));
+        if (need_finish_sorting)
         {
-            auto transform = std::make_shared<MergingSortedTransform>(
-                    pipeline.getHeader(),
-                    pipeline.getNumStreams(),
-                    result_description,
-                    max_block_size,
-                    SortingQueueStrategy::Batch,
-                    limit);
-
-            pipeline.addTransform(std::move(transform));
+            finishSorting(pipeline, prefix_description, result_description, limit);
         }
+        return;
     }
-    else if (type == Type::MergingSorted)
-    {        /// If there are several streams, then we merge them into one
-        if (pipeline.getNumStreams() > 1)
+
+    const auto input_sort_mode = input_streams.front().sort_mode;
+    const SortDescription & input_sort_desc = input_streams.front().sort_description;
+    if (optimize_sorting_for_input_stream)
+    {
+        /// skip sorting if stream is already sorted
+        if (input_sort_mode == DataStream::SortMode::Stream && input_sort_desc.hasPrefix(result_description))
+            return;
+
+        /// merge sorted
+        if (input_sort_mode == DataStream::SortMode::Port && input_sort_desc.hasPrefix(result_description))
         {
-            auto transform = std::make_shared<MergingSortedTransform>(
-                    pipeline.getHeader(),
-                    pipeline.getNumStreams(),
-                    result_description,
-                    max_block_size,
-                    SortingQueueStrategy::Batch,
-                    limit);
-
-            pipeline.addTransform(std::move(transform));
+            type = Type::MergingSorted;
```

Comment:
Right now we never call transformPipeline more than once, but there may be good to allow it for some optimizations.
Let's try not to change internal state here.
(Maybe I should have done transformPipeline const)